Don't expand CIDR labels, match smartly in Labels instead

squeed opened the pull request:
CIDR labels are expensive. Very expensive. For an IPv6 prefix, `GetCidrLabels()` can take over 2 KiB of memory.

This change refactors the label internals so that CIDR labels match when the cidr is larger as well. Then, we no longer have to expand CIDR labels at all; the matching logic handles it correctly.

Note to reviewers: Please review by commit; earlier commits add some test invariants. Another commit lightly refactors the label matching API to try and bring some sanity. Only the final two commits change any logic.

squeed committed:
labels: don't expand CIDR labels

Previously, we would expand a CIDR in to the full set of possible CIDRs
that could select it. For example "1.1.1.1/32" would be expanded in to
[0.0.0.0/0, 0.0.0.0/1, ... 1.1.1.0/31, 1.1.1.0/32]. This causes
significant memory and CPUusage, especially for circumstances such as ToFQDN
policies where many /32 and /128 identities are created.

Now that CIDR selectors are prefix-aware, rather than just string
matches, we can stop generating the complete list of CIDRs. This is safe
because CIDRs labels now select CIDRs that are contained within.

Benchmark results:

                     │ ../bench_main.out │         ../bench_cidr.out
                     │      sec/op       │   sec/op    vs base
UpdateGenerateDNS-12          4.972 ± 2%   2.882 ± 3%  -42.02% (p=0.000 n=10)

                     │ ../bench_main.out │          ../bench_cidr.out
                     │       B/op        │     B/op      vs base
UpdateGenerateDNS-12        77.26Mi ± 0%   24.52Mi ± 0%  -68.26% (p=0.000 n=10)

                     │ ../bench_main.out │          ../bench_cidr.out
                     │     allocs/op     │  allocs/op   vs base
UpdateGenerateDNS-12         508.0k ± 0%   291.7k ± 0%  -42.59% (p=0.000 n

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

joestringer commented on pkg/policy/selectorcache_selector.go:
```diff
@@ -182,25 +182,13 @@ func (f *fqdnSelector) setSelectorIPs(ips []netip.Addr) {
 // that is in wantLabels.
 // This is reasonably efficient, as it relies on both arrays being sorted.
 func (f *fqdnSelector) matches(identity scIdentity) bool {
-	wantIdx := 0
```
Is the function comment here still accurate with the new implementation?

joestringer commented on pkg/policy/selectorcache_selector.go:
```diff
@@ -182,25 +182,13 @@ func (f *fqdnSelector) setSelectorIPs(ips []netip.Addr) {
-			wantIdx++
+	for _, wantLbl := range f.wantLabels {
+		for _, haveLbl := range identity.lbls {
+			if wantLbl.Matches(&haveLbl) {
```
nit, Mildly curious whether there's anything to look into here in that until now `Label` didn't have a public `Matches` but we always used `LabelArray` and its corresponding `Has` / `Contains` / `Lacks` APIs. Not a big deal but maybe rather than iterating `identity.lbls` it could/should just use `identity.Has(...)`...?

squeed replied:
I added a new method, `Intersects()`, that does the same thing. Perhaps it's useful. Otherwise, yes, changing this to `.Has()` would be the right idea.

joestringer commented on pkg/labels/array.go:
```diff
@@ -113,24 +113,17 @@ nextLabel:
 func (ls LabelArray) Has(key string) bool {
 	// The key is submitted in the form of `source.key=value`
 	keyLabel := parseSelectLabel(key, '.')
-	if keyLabel.IsAnySource() {
```
Do we have good unit tests for the "any source" handling of this? Combinations of selectors selecting "any" source and matching keys, or selectors with specific sources and matching/nonmatching sources for the key?

Seeing the code for source handling getting dropped in a commit that is specifically about one special source and its own keys makes me a little bit nervous.

squeed replied:
We do now; I added some tests *first*, then refactored. So we should be good here.

joestringer commented on pkg/labels/labels.go:
```diff
@@ -267,9 +267,45 @@ func (l *Label) IsReservedSource() bool {
+
+	// Do cidr-aware matching if both sources are "cidr".
+	// This means that "cidr:10.0.0.0/24".matchesKey("cidr:10.0.0.1/32") = true
+	// Note that this does not take effect if either label has an "any" key.
```
I don't quite understand the label to label level here but I would think that a selector for `"any:10.0.0.0/24".matchesKey("cidr:10.0.0.1/32") == true`. Source any at the selector level is ambivalent about the source of the actual label it's targeting. Little hard to see through the abstractions to here given this is on a label-by-label basis rather than selector behaviour for some labels, but this surprises me a bit.

squeed replied:
I'm concerned about accidentally matching CIDRs for the `any:` case. CIDR parsing is somewhat lenient; I want to play around and see if there are any textual labels that might trigger an IPv6 prefix. That would be a rude surprise.

joestringer replied:
Is this only restricting source checking for CIDR labels or for all labels? There is a small chance that someone, somewhere, relies on the current `any` behavior. Probably not for CIDRs specifically - I'd expect that the policy parsing for ToCIDR statements is sanitizing this already. But maybe for other labels.

squeed replied:
Just cidr behavior.

FWIW, “any” selection is used extensively already by the policy engine. So I had broken that here, all the tests would be broken.

squeed replied:
Just to close the loop; there are now some any-touching test cases. And, as said before, we actually use "any" extensively.

The only open question is if we should parse `any:cidr` labels. I feel like no; ToEndpoints selectors should not select CIDRs.

joestringer reviewed (commented):
In principle this seems like a really neat win. I can't currently think of any issues with encoding this into the selectors to identify labels with prefixes that are more specific than the selector.

I didn't review in deep detail beyond the second patch, primarily just for time and because I think you were interested in high level feedback at this time. Perhaps another @cilium/sig-policy member could do further indepth review or I can reschedule a subsequent follow-up review later on.

squeed committed:
labels / policy: have tests mirror actual usage of LabelSourceAny

We actually rely quite heavliy on the LabelSourceAny mechanism.
EndpointSelectors in CiliumNetworkPolicies always have LabelSourceAny
added.

For example, the block
```yaml
toEndpoints:
  - matchLabels:
      io.kubernetes.pod.namespace: kube-system
      k8s-app: kube-dns
```

converts to the label selector

`{any.io.kubernetes.pod.namespace: kube-system,any.k8s-app: kube-dns,}`

So, explicitly mention this in comments and update the SelectorCache
tests to capture this behavior.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy: small cidr selector test cases

Adding some invariants that should not be broken during coming
refactors.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
pkg/labels: add simple CIDR test

This tests adds a very specific invariant that is needed by the policy
engine. Specifically, the expanded set of CIDR labels must always
`.Has()` a CIDR label that contains it.

This will be relevant when we stop expanding CIDR labels and, instead,
logically compute CIDR matching.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
labels: clean up label matching API

The label matching API is complicated and inconsistent. This change
tries to bring some sanity to the API going forward, without changing
existing behavior.

Label matching is directional / non-communtative. Specifically,
`"any:foo=bar".Equals("k8s:foo=bar")` is true, whereas
`"k8s:foo=bar".Equals("any:foo=bar")` is false.

So, with the eventual goal of removing `Label.Equals()`, this commit
adds a new `Label.Has()` and `Label.HasKey()` api, with clear
documentation around directionality.

The fixed point here is `LabelArray.Has()`, which needs a specific
directionality as required by the k8s label selector library. Everything
else is based off of that.

This also changes `Labels.Has()` to match directionality w.r.t
`any`-source selectors. In theory this is a breaking change, in
actuality `Labels.Has()` is never passed `any` selectors, so this is
moot.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
labels: CIDR-aware label selection

This changes the Labels API to be CIDR-aware. It then logically
"expands" CIDR labels when computing matches, so that selectors can
match CIDRs even when not present. It does this by parsing CIDRs on
label creation, then checking CIDR overlap in the `MatchesKey()`
function.

The API contract we expose to the policy engine is unchanged:
```
GetCIDRLabels("10.0.0.0/24").LabelArray().Has("cidr.10.0.0.0/8") == true
```

The goal is to stop manually expanding CIDR labels, which is very
inefficient. This will follow in a subsequent commit.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed commented on pkg/labels/array.go:
```diff
@@ -96,13 +96,41 @@ nextLabel:
+		}
+
+		// Not equal, bump
+		if needed[n].Key < ls[l].Key {
```
I'm pretty sure this is **wrong wrong wrong** now, since sorting CIDR strings alphabetically does not necessarily match ordering. It was a cute trick, but a bad idea.

squeed replied:
Yup, totally wrong! `["cidr:11.11.11.11/32"].Intersects(["cidr:110.0.0.0/8", "cidr:8.0.0.0/5"])` should be true, but it wasn't.

squeed replied:
(fixed, test case added, but leaving for visibility)

nathanjsweet commented on pkg/labels/labels.go:
```diff
@@ -252,9 +252,21 @@ func (l *Label) IsReservedSource() bool {
+// MatchesKey returns true if l has target's key.
+// note that l may be looser than target.
+func (l *Label) MatchesKey(target *Label) bool {
+	if !l.IsAnySource() && l.Source != target.Source {
```
> Label matching is directional / non-communtative. Specifically,
`"any:foo=bar".Equals("k8s:foo=bar")` is true, whereas
`"k8s:foo=bar".Equals("any:foo=bar")` is false.

What about `"any:foo=bar".Equals("any:foo=bar")` ? That probably never happens though :thinking: .

nathanjsweet commented on pkg/labels/labels.go:
```diff
@@ -143,6 +147,10 @@ type Label struct {
+
+	// optimization for CIDR prefixes
+	// +deepequal-gen=false
+	cidr *netip.Prefix `json:"-"`
```
Should we lazy eval here and turn `cidr` into a private method rather than a member to always be pre-calculated. The cidr label will not always be compared or parsed. Thoughts?

squeed replied:
I looked in to this, and it is probably better not to lazy eval. Every time we add a CIDR identity, we do (pseudocode)

```
for selector in selectors: (N = 100)
  for identity in identities: (N = 1000)
    for label in identity: (N = 2)
      if label.Has(Selector):
        // ...
```

means there is definitely a cost to computing the CIDR every time. Plus, doing so requires heap allocation, since there are string substitutions. Meaning we would throw off a lot of garbage.

nathanjsweet commented on pkg/labels/labels.go:
```diff
@@ -227,11 +235,21 @@ func NewLabel(key string, value string, source string) Label {
 		Value:  value,
 		Source: source,
 	}
+	if l.Source == LabelSourceCIDR {
```
See lazy eval comment.

nathanjsweet reviewed (changes requested):
Very impressive! Thanks! Two questions.

christarazi commented on pkg/labels/array.go:
```diff
@@ -96,13 +96,29 @@ nextLabel:
+// This has the same matching semantics as Has, namely,
+// ["k8s:foo=bar"].Intersects(["any:foo=bar"]) == true
+// ["any:foo=bar"].Intersects(["k8s:foo=bar"]) == false
+func (ls LabelArray) Intersects(needed LabelArray) bool {
```
Should there be a mention here that it's desirable for the label arrays to be sorted to take advantage of efficiency and warn the caller that if not sorted, then it can be expensive?

squeed replied:
(LabelArrays are required to be sorted, so that's not an issue).

The sorting requirement is now moot anyways, since it doesn't work for CIDR expansion :-/. Though that does make me want to be clever and use binary search in `.Has()` now... Task for another PR.

christarazi commented on pkg/labels/array.go:
```diff
@@ -124,18 +140,11 @@ nextLabel:
-			}
+	for _, l := range ls {
+		// note inverted order; by definition, key is "looser" as it
+		// is coming from a selector.
```
I'm not sure I'm following what's "inverted" here?

christarazi replied:
I'm also not following what "key is looser" means?

squeed replied:
I'm trying to find a way to express in which direction source `any` is respected. Thoughts? This is just a small comment for the next viewer.

Maybe I should replace `Matches()` and `MatchesKey()` with just `Has(key Label)`, keeping the ordering consistent.

squeed replied:
OK, I made `Label.Has()` and `Label.HasKey()` that works the same way as `LabelArray.Has()` and `Labels.Has()`. Now everything is a lot more consistent :relieved:.

christarazi commented on pkg/labels/cidr.go:
```diff
@@ -78,6 +78,16 @@ func IPStringToLabel(ip string) (Label, error) {
+		return netip.Prefix{}, fmt.Errorf("failed to parse label prefix %s: %w", key, err)
+	}
+	return pfx, nil
+
```
Nit: newline

christarazi commented on pkg/labels/labels.go:
```diff
@@ -227,11 +235,21 @@ func NewLabel(key string, value string, source string) Label {
+	if l.Source == LabelSourceCIDR {
+		c, err := LabelToPrefix(l.Key)
+		if err != nil {
+			logrus.WithField("key", l.Key).WithError(err).Error("Failed to parse CIDR label: invalid prefix.")
```
I understand the intent to log errors here if one occurs, but if we think about where we are in the stack, does it really make sense to log in `pkg/labels/labels.go`? My instinct tells me that if we have a CIDR label then it's not going to be malformed by this point and if it is, we should investigate higher up in the stack where that can happen and prevent it from "leaking" down here. 

The problem with logging an error here is that it's not clear what the user should even do. It's a very low-level detail error log.

squeed replied:
Really good perspective. I'm not entirely sure what the right course of action is here.

CIDR labels really cannot be malformed; everything that parses them already has proper validation somewhere higher up the stack. I added this error message here just for insurance that we will fail CI if we somehow enter this case.

One more insane option would be to print a stacktrace here (but not panic) if we reach this point. At least in debug mode.

joestringer replied:
This seems like something that should be sanitized much much earlier, then the action item for the user is "Report this to the developers". Ideally no-one ever sees the error, unless we somehow missed some sanitization somewhere.

Better still is to avoid passing strings around internally and instead pass around the type after it was sanitized, but I don't know if that's particularly viable. Maybe if `Label` was an interface and we had typed labels that are either string or `netip.Prefix`...? Getting a bit speculative here.

squeed replied:
We're stuck parsing strings, since the API *has* to be `LabelArray.Has("cidr.10.0.0.0/8")` as required by the k8s label selector package. So, this warning - which I never expect to happen - has to be here.

christarazi replied:
If we can assume that by the time we reach this code it's impossible to fail to parse and the risk for panic is 0, then let's turn the `netip` call to `MustParsePrefix` which doesn't return an error and therefore, remove the log.

squeed replied:
MustParsePrefix panics on invalid input; I really don’t want that to happen on the off chance someone manages to slip a bad policy in.

No great answer here :-/

christarazi commented on pkg/labels/cidr_test.go:
We're removing a whole bunch of benchmarks here but do we understand the performance impact of this commit? How much of an impact does it make?

squeed replied:
Those benchmarks were testing how quickly we could generate the 1000s of labels necessary for under 50 IPs, and to prove that the LRU cache worked. We now generate 1 label and need no caching, so the benchmarks are somewhat irrelevant. I can keep them if you like, but they're not super interesting.

christarazi replied:
Makes sense, no need to keep them then :broom:

pippolo84 commented on pkg/labels/array_test.go:
```diff
@@ -293,3 +295,34 @@ func BenchmarkLabelArray_String(b *testing.B) {
+		NewLabel("foo", "bar", "k8s"),
+		NewLabel("foo1", "bar1", "any"), // not valid, but good to capture
+		NewLabel("kube-apiserver", "", "reserved"),
+		NewLabel("2001-db8-cafe--0/54", "", "cidr"),
```
Nit: isn't this one already added as part of `GetCIDRLabels(netip.MustParsePrefix("2001:db8:cafe::/54"))`?

pippolo84 reviewed (approved):
Great, thanks!
Just a very minor nit regarding a test.

christarazi commented on pkg/labels/cidr.go:
(Leaving a comment here to have a discussion point.)

squeed replied:
I have a pprof lying around from a user cluster that shows significant memory cost specifically from allocating CIDR identities. Additionally, it shows a lot (30%) of CPU time dedicated to GC when the FQDN subsystem is under high churn.

This is a fair question, though; I think there's a reasonable benchmark in the code that can show the benefit. Let me see what I can actually demonstrate.

squeed replied:
I whipped up a quick benchmark, the results are promising :-)

```
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/daemon/cmd
cpu: 12th Gen Intel(R) Core(TM) i7-1250U
                     │ ../bench_main.out │         ../bench_cidr.out          │
                     │      sec/op       │   sec/op    vs base                │
UpdateGenerateDNS-12          4.972 ± 2%   2.882 ± 3%  -42.02% (p=0.000 n=10)

                     │ ../bench_main.out │          ../bench_cidr.out           │
                     │       B/op        │     B/op      vs base                │
UpdateGenerateDNS-12        77.26Mi ± 0%   24.52Mi ± 0%  -68.26% (p=0.000 n=10)

                     │ ../bench_main.out │          ../bench_cidr.out          │
                     │     allocs/op     │  allocs/op   vs base                │
UpdateGenerateDNS-12         508.0k ± 0%   291.7k ± 0%  -42.59% (p=0.000 n=10)
```

christarazi replied:
Looks great! Could we add the benchmark results to the commit msgs?

christarazi reviewed (approved):
The changes LGTM and I can't think of any adverse effects from the PR. I am curious though if we have an understanding of what the actual performance impact is with this change. AFAIU, `GetCIDRLabels` consumed a lot of CPU and memory to generate CIDR labels constantly when the DNS proxy sees many DNS requests, but how much has this PR really changed? Have we done a pprof comparison, for example? I don't doubt that this PR improved things, but it's also useful to quantify how much has changed and worst case, potentially point out a flaw that we didn't detect previously.
